### ISSUE TITLE
java.io.IOException when running

### ISSUE THREAD

**Issue by jdraffin** [agclark12/autocrop_black_edges#3]
Hi Andrew. Thanks for your work on the plugin --- definitely just what I need and adds a sorely needed functionality to otherwise excellent registration plugins (I use Correct 3D Drift).

I also get an exception while running the plugin, both for the sample stack and my own (replicated on a colleague's machine in the latter case). I've created a new issue here because the exception seems to be different from the one halloleo received. I'm using ImageJ 1.54d (Fiji distribution) with Java 1.8.0_341 (64b-bit), if that might be relevant. The console output reads as follows (same for sample stack and mine):

> [ERROR] java.io.IOException: Mark invalid
> 	at java.io.BufferedReader.reset(BufferedReader.java:512)
> 	at org.python.core.ParserFacade.parseExpressionOrModule(ParserFacade.java:133)
> 	at org.python.util.PythonInterpreter.compile(PythonInterpreter.java:321)
> 	at org.python.util.PythonInterpreter.compile(PythonInterpreter.java:317)
> 	at org.python.jsr223.PyScriptEngine.compileScript(PyScriptEngine.java:90)
> 	at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:31)
> 	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
> 	at org.scijava.script.ScriptModule.run(ScriptModule.java:164)
> 	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:63)
> 	at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:225)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:750)
> java.io.IOException: java.io.IOException: Mark invalid
> 
> 	at org.python.core.Py.JavaError(Py.java:547)
> 	at org.python.core.ParserFacade.fixParseError(ParserFacade.java:107)
> 	at org.python.core.ParserFacade.parseExpressionOrModule(ParserFacade.java:136)
> 	at org.python.util.PythonInterpreter.compile(PythonInterpreter.java:321)
> 	at org.python.util.PythonInterpreter.compile(PythonInterpreter.java:317)
> 	at org.python.jsr223.PyScriptEngine.compileScript(PyScriptEngine.java:90)
> 	at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:31)
> 	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
> 	at org.scijava.script.ScriptModule.run(ScriptModule.java:164)
> 	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:63)
> 	at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:225)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:750)
> Caused by: java.io.IOException: Mark invalid
> 	at java.io.BufferedReader.reset(BufferedReader.java:512)
> 	at org.python.core.ParserFacade.parseExpressionOrModule(ParserFacade.java:133)
> 	... 14 more

Any idea what might be going on?

Cheers,
Jonny

**Comment by agclark12** [comment]
Hi Johnny, thanks for the feedback. Sorry you're getting an error here. I might not have a chance to look into this until next week. At the end of the stack trace it says "... 14 more". Can you get those last 14 lines of the error?

**Comment by jdraffin** [comment]
Thanks for the reply Andrew.

I'm afraid the console output is truncated and "... 14 more" isn't clickable. I scoured ImageJ for any relevant options relating to limiting console output but didn't come up with anything.

I did try running the script in an installation (1.53k) of plain ImageJ (after updating Jython) and got the following in the log:

`SyntaxError: ("no viable alternative at character '\xc2'", ('<string>', 72, 62, ''))
`

**Comment by jdraffin** [comment]
Oh dear.

I originally downloaded the script with "Save Link As" and it saved as a .py file, but I've just opened it and found html code inside! Replacing this with your python code and it works like a charm with the sample stack --- the not-so-wild goose is back in its pen. Bit of a github lesson for me --- sorry for the false alarm.

It also works perfectly with 4D stacks of mine (xyzc), which will be really useful, although not with 4D or 5D hyperstacks (xyzt or xyztc). This would certainly be a nice feature, but with a bit of macro code the plugin can do its heavy lifting on stacks that are then concatenated back into a hyperstack. Thanks again for the really useful contribution and attempting to solve a problem of my own making.

**Comment by agclark12** [comment]
Well, I'm happy the problem has been solved. It would have probably been tough for me to debug this one I guess. :) Anyway, glad to hear that it works with your stacks now. I wrote this with 3D image stacks in mind (usually xyt, but xyz should also work). Adding more dimensions would of course be possible; maybe I will try to implement this at some point. Could quickly get more complicated because maybe some users may want to choose what dimension(s) to consder for the cropping... If you just want the smallest rectangle across all dimensions, probably the easiest thing to do would be to convert the hyperstack to a flat stack, then run the macro, then reconvert back to a hyperstack. But maybe that's what you mean above. In any case, glad to hear that it's useful for you!
